File: app/src/test/java/com/metahumanz/pacilread/reader/ReplacementEngineTest.java
package com.metahumanz.pacilread.reader;

import com.metahumanz.pacilread.model.ReplacementRuleRecord;

import org.junit.Test;

import java.lang.reflect.Field;
import java.util.Collections;
import java.util.Map;
import java.util.regex.Pattern;

import static org.junit.Assert.assertEquals;

public class ReplacementEngineTest {
    @Test
    public void regexRulesReuseCompiledPatternUntilRuleChanges() throws Exception {
        Map<String, Pattern> cache = compiledRegexCache();
        synchronized (cache) {
            cache.clear();
        }

        ReplacementRuleRecord rule = regexRule(1L, "(\\d+)", "[$1]", 100L);

        assertEquals("a[123]", ReplacementEngine.apply("a123", Collections.singletonList(rule)));
        synchronized (cache) {
            assertEquals(1, cache.size());
        }

        assertEquals("b[456]", ReplacementEngine.apply("b456", Collections.singletonList(rule)));
        synchronized (cache) {
            assertEquals(1, cache.size());
        }

        rule.updatedAt = 101L;
        assertEquals("c[789]", ReplacementEngine.apply("c789", Collections.singletonList(rule)));
        synchronized (cache) {
            assertEquals(2, cache.size());
        }
    }

    @Test
    public void literalRulesDoNotPopulateRegexCache() throws Exception {
        Map<String, Pattern> cache = compiledRegexCache();
        synchronized (cache) {
            cache.clear();
        }

        ReplacementRuleRecord rule = new ReplacementRuleRecord();
        rule.id = 2L;
        rule.pattern = "foo";
        rule.replacement = "bar";
        rule.active = true;
        rule.regex = false;

        assertEquals("bar baz", ReplacementEngine.apply("foo baz", Collections.singletonList(rule)));
        synchronized (cache) {
            assertEquals(0, cache.size());
        }
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Pattern> compiledRegexCache() throws Exception {
        Field field = ReplacementEngine.class.getDeclaredField("COMPILED_REGEX_CACHE");
        field.setAccessible(true);
        return (Map<String, Pattern>) field.get(null);
    }

    private static ReplacementRuleRecord regexRule(long id, String pattern, String replacement, long updatedAt) {
        ReplacementRuleRecord rule = new ReplacementRuleRecord();
        rule.id = id;
        rule.pattern = pattern;
        rule.replacement = replacement;
        rule.active = true;
        rule.regex = true;
        rule.updatedAt = updatedAt;
        return rule;
    }
}

File: app/src/test/kotlin/com/metahumanz/pacilread/reader/ReplacementEngineTest.kt
package com.metahumanz.pacilread.reader

import com.metahumanz.pacilread.model.ReplacementRuleRecord
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.regex.Pattern

class ReplacementEngineTest {
    @Test
    fun regexRulesReuseCompiledPatternUntilRuleChanges() {
        val cache = compiledRegexCache()
        synchronized(cache) { cache.clear() }
        val rule = regexRule(1L, "(\\d+)", "[$1]", 100L)

        assertEquals("a[123]", ReplacementEngine.apply("a123", listOf(rule)))
        synchronized(cache) { assertEquals(1, cache.size) }
        assertEquals("b[456]", ReplacementEngine.apply("b456", listOf(rule)))
        synchronized(cache) { assertEquals(1, cache.size) }

        rule.updatedAt = 101L
        assertEquals("c[789]", ReplacementEngine.apply("c789", listOf(rule)))
        synchronized(cache) { assertEquals(2, cache.size) }
    }

    @Test
    fun literalRulesDoNotPopulateRegexCache() {
        val cache = compiledRegexCache()
        synchronized(cache) { cache.clear() }
        val rule = ReplacementRuleRecord().apply {
            id = 2L
            pattern = "foo"
            replacement = "bar"
            active = true
            regex = false
        }
        assertEquals("bar baz", ReplacementEngine.apply("foo baz", listOf(rule)))
        synchronized(cache) { assertEquals(0, cache.size) }
    }

    @Suppress("UNCHECKED_CAST")
    private fun compiledRegexCache(): MutableMap<String, Pattern> {
        val field = ReplacementEngine::class.java.getDeclaredField("COMPILED_REGEX_CACHE")
        field.isAccessible = true
        return field.get(null) as MutableMap<String, Pattern>
    }

    private fun regexRule(id: Long, pattern: String, replacement: String, updatedAt: Long) =
        ReplacementRuleRecord().apply {
            this.id = id
            this.pattern = pattern
            this.replacement = replacement
            active = true
            regex = true
            this.updatedAt = updatedAt
        }
}
